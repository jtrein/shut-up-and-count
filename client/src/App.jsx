import React from 'react';

import logo from './logo.svg';
import styles from './App.styles';
import commonStyles from './common.styles';
import BigButton from './BigButton';

const App = () => (
  <div style={styles.wrap}>
    <header className="App-header">
      <img src={logo} className="App-logo" alt="logo" />
      <h1 style={commonStyles.hide}>Shut Up and Count</h1>
    </header>

    <BigButton disabled={false} onClick={() => {}} />
  </div>
);

export default App;
