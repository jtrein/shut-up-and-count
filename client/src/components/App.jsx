import React, { Component } from 'react';
import Spinner from 'react-spinkit';
import { Column, Table } from 'react-virtualized';
import Snackbar from 'material-ui/Snackbar';
import 'react-virtualized/styles.css';

import { MSG_SERVER_ERR } from '../settings';
import postTextFile from '../api';
import BigButton from './BigButton';
import FileDropzone from './FileDropzone';
import Header from './Header';
import styles from '../styles/App.styles';

const rowGetter = (index, wordsList) => ({
  word: wordsList[index][0],
  occurrences: wordsList[index][1],
});

class App extends Component {
  constructor() {
    super();

    this.state = {
      data: {},
      error: '',
      fileToSend: null,
      isButtonDisabled: true,
      progress: false,
    };
  }

  handleDropAccepted(file) {
    this.setState({ isButtonDisabled: false, fileToSend: file });
  }

  handleHideError() {
    this.setState({ error: '' });
  }

  handlePostFile() {
    if (!this.state.fileToSend) return;

    this.setState({ progress: true });

    postTextFile(this.state.fileToSend)
      .then((data) => {
        if (data.error) {
          throw Error(data.error);
        }

        this.setState({ data: data || {}, progress: false });
      })
      .catch((e) => {
        this.setState({ error: e.message || MSG_SERVER_ERR, progress: false });
      });
  }

  render() {
    const {
      data,
      error,
      isButtonDisabled,
      progress,
    } = this.state;
    const { eachWordCount } = data;
    const wordsList = eachWordCount ? Object.entries(eachWordCount) : [];

    return (
      <div style={styles.wrap}>
        <Header />

        <FileDropzone onDropAccepted={file => this.handleDropAccepted(file)} />

        {/* upload & parse */}
        <BigButton disabled={isButtonDisabled} onClick={() => this.handlePostFile()} />

        {/* progress */}
        { progress && !error && <Spinner name="triangle-skew-spin" color="#662e91" /> }

        {
          wordsList.length ? (
            <div style={styles.tableWrap}>
              <p style={styles.totalWordCount}>
                Total word count: <b>{data.totalWordCount}</b>
              </p>

              <Table
                width={500}
                height={500}
                headerHeight={20}
                rowHeight={30}
                rowCount={wordsList.length}
                rowGetter={({ index }) => rowGetter(index, wordsList)}
                rowStyle={styles.tableRow}
              >
                <Column
                  label="Word"
                  dataKey="word"
                  width={450}
                />
                <Column
                  width={200}
                  label="Occurrences"
                  dataKey="occurrences"
                />
              </Table>
            </div>
          ) : null
        }

        {/* server error notify */}
        <Snackbar
          autoHideDuration={3000}
          open={Boolean(this.state.error)}
          onClose={() => this.handleHideError()}
          message={this.state.error}
        />
      </div>
    );
  }
}

export default App;
