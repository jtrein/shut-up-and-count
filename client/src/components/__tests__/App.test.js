/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Snackbar from 'material-ui/Snackbar';

import MockFetch from '../../__mocks__/fetch';
import App from '../App';

Enzyme.configure({ adapter: new Adapter() });

test('renders without crashing', () => {
  const div = document.createElement('div');
  render(<App />, div);
  unmountComponentAtNode(div);
});

test('renders expected initial components', () => {
  const wrapper = shallow(<App />);
  expect(wrapper.find('Header').length).toEqual(1);
  expect(wrapper.find('FileDropzone').length).toEqual(1);
  expect(wrapper.find('BigButton').length).toEqual(1);
  expect(wrapper.find('Snackbar').length).toEqual(0);
  expect(wrapper.find('Table').length).toEqual(0);
  expect(wrapper.find('Column').length).toEqual(0);
  expect(wrapper.find('Spinner').length).toEqual(0);
});

test('BigButton is disabled on init', () => {
  const wrapper = shallow(<App />);
  expect(wrapper.find('BigButton').props().disabled).toEqual(true);
});

test('BigButton can enable when isButtonDisabled=false', () => {
  const wrapper = shallow(<App />);
  wrapper.setState({ isButtonDisabled: false });
  expect(wrapper.find('BigButton').props().disabled).toEqual(false);
});

test('BigButton click: spinner does not fire in initial state', () => {
  const wrapper = shallow(<App />);
  wrapper.find('BigButton').simulate('click');
  expect(wrapper.state().progress).toEqual(false);
});

test('BigButton click: postTextFile sets state after fetch', async () => {
  const origFetch = global.fetch;
  const fakeRes = [{ eachWordCount: { hello: 1 }, totalWordCount: 1 }];
  global.fetch = MockFetch(fakeRes);

  const wrapper = shallow(<App />);
  wrapper.setState({ fileToSend: [{ filename: 'hello' }] });
  wrapper.find('BigButton').simulate('click');

  expect(wrapper.state().progress).toEqual(true);

  await wrapper;
  await wrapper;

  expect(wrapper.state().progress).toEqual(false);
  expect(wrapper.state().data).toEqual(fakeRes);

  global.fetch = origFetch;
});

test('BigButton click: postTextFile sets error state after fetch', async () => {
  const origFetch = global.fetch;
  const fakeRes = { error: 'is this bad?' };
  global.fetch = MockFetch(fakeRes);

  const wrapper = shallow(<App />);
  wrapper.setState({ fileToSend: [{ filename: 'hello' }] });
  wrapper.find('BigButton').simulate('click');

  await wrapper;
  await wrapper;
  await wrapper;

  expect(wrapper.state().data).toEqual({});
  expect(wrapper.state().error).toEqual('is this bad?');
  expect(wrapper.state().progress).toEqual(false);

  global.fetch = origFetch;
});

test('Snackbar shows when error is set', () => {
  const wrapper = shallow(<App />);
  wrapper.setState({ error: 'hello' });
  expect(wrapper.find(Snackbar).length).toEqual(1);
});

test('Spinner can show when progress=true and no error', () => {
  const wrapper = shallow(<App />);
  wrapper.setState({ progress: true });
  expect(wrapper.find('Spinner').length).toEqual(1);
});

test('Table can show when data', () => {
  const dataFixture = { eachWordCount: { hello: 2, world: 4 } };
  const wrapper = shallow(<App />);
  wrapper.setState({ data: dataFixture });
  expect(wrapper.find('Table').length).toEqual(1);
});

test('Table can show <Column /> when data', () => {
  const dataFixture = { eachWordCount: { hello: 2 } };
  const wrapper = shallow(<App />);
  wrapper.setState({ data: dataFixture });
  expect(wrapper.find('Column').length).toEqual(2);
});

test('Table can show text before table', () => {
  const dataFixture = { eachWordCount: { hello: 2 }, totalWordCount: 1 };
  const wrapper = shallow(<App />);
  wrapper.setState({ data: dataFixture });
  expect(wrapper.find('p').length).toEqual(1);
  expect(wrapper.find('p').text()).toEqual('Total word count: 1');
});

test('Table can show text before table', () => {
  const dataFixture = { eachWordCount: { hello: 2 }, totalWordCount: 1 };
  const wrapper = mount(<App />);
  wrapper.setState({ data: dataFixture });
  expect(wrapper.find('Table').text()).toContain('Word');
  expect(wrapper.find('Table').text()).toContain('Occurrences');
  expect(wrapper.find('Table').text()).toContain('hello');
  expect(wrapper.find('Table').text()).toContain('2');
});

test('handleDropAccepted can set proper state', () => {
  const wrapper = mount(<App />);
  wrapper.find('FileDropzone').simulate('drop');
  expect(wrapper.state().fileToSend).toEqual([]);
  expect(wrapper.state().isButtonDisabled).toBe(false);
});
