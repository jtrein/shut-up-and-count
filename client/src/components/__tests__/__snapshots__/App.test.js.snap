// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: renders as expected 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "bottom": 0,
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "center",
      "left": 0,
      "marginTop": 24,
      "right": 0,
      "top": 0,
    }
  }
>
  <Header />
  <FileDropzone
    onDropAccepted={[Function]}
  />
  <BigButton
    disabled={true}
    onClick={[Function]}
  />
  <div
    id="table"
    style={
      Object {
        "marginTop": 48,
      }
    }
  />
  <WithStyles(Snackbar)
    autoHideDuration={3000}
    message=""
    onClose={[Function]}
    open={false}
  />
</div>
`;
